File: src/shared/ui/menu-anchor.ts
/**
 * Where a menu is drawn once it no longer lives inside its card.
 *
 * A menu positioned `absolute` inside a card is painted with that card, so a
 * later sibling in the grid paints OVER it — reported on the Develop gallery,
 * where the ⋯ menu of a roll slid under the roll beneath it — and the page's
 * own scroll container clips it at the bottom row. The cure is to draw it in a
 * portal at fixed viewport coordinates, which is what this module computes:
 * given the trigger's rect and the menu's measured size, the corner it opens
 * from, the height it may take, and the side it ended up on.
 *
 * `side` and `align` are therefore a PREFERENCE, not a placement: a menu that
 * would fall off the bottom flips above, and one that would fall off an edge
 * slides back in. A menu that fits nowhere keeps `MIN_ROOM` and scrolls —
 * better a short scrolling menu than one drawn past the screen.
 *
 * DOM-free and tested: nothing here reads an element, the caller measures.
 */

export interface AnchorRect {
  left: number;
  right: number;
  top: number;
  bottom: number;
}

export interface MenuAnchorInput {
  /** The trigger, in viewport coordinates (`getBoundingClientRect`). */
  trigger: AnchorRect;
  /** The menu's natural size, measured unconstrained. */
  menu: { width: number; height: number };
  viewport: { width: number; height: number };
  /** Preferred side; flipped when it does not fit. */
  side: 'below' | 'above';
  /** Which of the menu's edges lines up with the trigger's. */
  align: 'end' | 'start';
  /** Between the trigger and the menu. */
  gap?: number;
  /** The least the menu keeps from a viewport edge. */
  margin?: number;
}

export interface MenuAnchor {
  left: number;
  top: number;
  maxHeight: number;
  /** Where it actually landed — the caller may want to know it flipped. */
  placed: 'below' | 'above';
}

/** Two items and a rule: below this a menu is not worth drawing, it scrolls. */
const MIN_ROOM = 96;

const clamp = (v: number, lo: number, hi: number) => (v < lo ? lo : v > hi ? hi : v);

export function menuAnchor({
  trigger,
  menu,
  viewport,
  side,
  align,
  gap = 6,
  margin = 8,
}: MenuAnchorInput): MenuAnchor {
  const roomBelow = viewport.height - margin - (trigger.bottom + gap);
  const roomAbove = trigger.top - gap - margin;

  // Flip only when the preferred side cannot hold the menu AND the other side
  // is better — a menu that fits nowhere stays where it was asked for.
  let placed = side;
  if (side === 'below' && menu.height > roomBelow && roomAbove > roomBelow) placed = 'above';
  if (side === 'above' && menu.height > roomAbove && roomBelow > roomAbove) placed = 'below';

  const room = placed === 'below' ? roomBelow : roomAbove;
  const maxHeight = clamp(Math.max(room, MIN_ROOM), 0, Math.max(viewport.height - 2 * margin, 0));
  const height = Math.min(menu.height, maxHeight);

  const wantedTop = placed === 'below' ? trigger.bottom + gap : trigger.top - gap - height;
  const top = clamp(wantedTop, margin, Math.max(margin, viewport.height - margin - height));

  const wantedLeft = align === 'end' ? trigger.right - menu.width : trigger.left;
  const maxLeft = viewport.width - margin - menu.width;
  const left = maxLeft < margin ? margin : clamp(wantedLeft, margin, maxLeft);

  return { left, top, maxHeight, placed };
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/UI/MenuAnchor.swift
// Where a menu is drawn once it no longer lives inside its card. Port of
// `src/shared/ui/menu-anchor.ts`.
//
// A menu positioned inside a card paints with that card, so a later sibling
// paints OVER it and the scroll container clips it at the bottom row; the cure
// is to draw it at fixed viewport coordinates, which is what this computes
// (`frontend.md`, «a menu is PORTALLED»). Given the trigger's rect and the
// menu's measured size: the corner it opens from, the height it may take,
// and the side it ended up on. `side` and `align` are a PREFERENCE, not a
// placement: a menu that would fall off the bottom flips above — only when the
// other side is BETTER, so a menu that fits nowhere stays where it was asked
// for and scrolls inside `minRoom`; one that would fall off an edge slides
// back in, and one wider than the screen gives up the margin rather than
// itself. Nothing here reads a view; the caller measures.

import Foundation

/// The trigger's edges in viewport coordinates (the web's
/// `getBoundingClientRect`), edges rather than an origin and a size.
public struct AnchorRect: Equatable, Sendable {
    public var left: Double
    public var right: Double
    public var top: Double
    public var bottom: Double

    public init(left: Double, right: Double, top: Double, bottom: Double) {
        self.left = left; self.right = right; self.top = top; self.bottom = bottom
    }
}

public enum MenuSide: String, Sendable {
    case below, above
}

/// Which of the menu's edges lines up with the trigger's.
public enum MenuAlign: String, Sendable {
    case end, start
}

public struct MenuAnchorInput: Equatable, Sendable {
    /// The trigger, in viewport coordinates.
    public var trigger: AnchorRect
    /// The menu's natural size, measured unconstrained.
    public var menu: Size
    public var viewport: Size
    /// Preferred side; flipped when it does not fit.
    public var side: MenuSide
    /// Which of the menu's edges lines up with the trigger's.
    public var align: MenuAlign
    /// Between the trigger and the menu.
    public var gap: Double
    /// The least the menu keeps from a viewport edge.
    public var margin: Double

    public init(trigger: AnchorRect, menu: Size, viewport: Size, side: MenuSide, align: MenuAlign,
                gap: Double = 6, margin: Double = 8) {
        self.trigger = trigger; self.menu = menu; self.viewport = viewport
        self.side = side; self.align = align; self.gap = gap; self.margin = margin
    }
}

public struct MenuAnchor: Equatable, Sendable {
    public var left: Double
    public var top: Double
    public var maxHeight: Double
    /// Where it actually landed — the caller may want to know it flipped.
    public var placed: MenuSide

    public init(left: Double, top: Double, maxHeight: Double, placed: MenuSide) {
        self.left = left; self.top = top; self.maxHeight = maxHeight; self.placed = placed
    }
}

/// Two items and a rule: below this a menu is not worth drawing, it scrolls.
public let menuMinRoom = 96.0

public func menuAnchor(_ input: MenuAnchorInput) -> MenuAnchor {
    let trigger = input.trigger
    let menu = input.menu
    let viewport = input.viewport
    let gap = input.gap
    let margin = input.margin

    let roomBelow = viewport.height - margin - (trigger.bottom + gap)
    let roomAbove = trigger.top - gap - margin

    // Flip only when the preferred side cannot hold the menu AND the other
    // side is better — a menu that fits nowhere stays where it was asked for.
    var placed = input.side
    if input.side == .below && menu.height > roomBelow && roomAbove > roomBelow { placed = .above }
    if input.side == .above && menu.height > roomAbove && roomBelow > roomAbove { placed = .below }

    let room = placed == .below ? roomBelow : roomAbove
    let tallest = max(viewport.height - 2 * margin, 0)
    let maxHeight = clamp(max(room, menuMinRoom), 0, tallest)
    let height = min(menu.height, maxHeight)

    let wantedTop = placed == .below ? trigger.bottom + gap : trigger.top - gap - height
    let lowestTop = max(margin, viewport.height - margin - height)
    let top = clamp(wantedTop, margin, lowestTop)

    let wantedLeft = input.align == .end ? trigger.right - menu.width : trigger.left
    let maxLeft = viewport.width - margin - menu.width
    let left = maxLeft < margin ? margin : clamp(wantedLeft, margin, maxLeft)

    return MenuAnchor(left: left, top: top, maxHeight: maxHeight, placed: placed)
}
